File: frontend/src/lib/types/randr.ts
export interface SchemeProgressKpis {
  total_affected_families: number;
  eligible_families: number;
  approved_families: number;
  allocated_families: number;
  completed_families: number;
  completion_percent: number;
}

export interface RAndRSchemeListItem {
  id: string;
  scheme_reference: string;
  scheme_title: string;
  scheme_type: string;
  project_id: string;
  project_title: string;
  project_code: string;
  sanctioned_budget_cr: number | string;
  spent_budget_cr: number | string;
  status: string;
  target_completion_date?: string | null;
  total_families_count: number;
  eligible_families_count: number;
  assisted_families_count: number;
  progress_percent: number;
  created_at: string;
}

export interface RAndRSchemeDetail extends RAndRSchemeListItem {
  state_name?: string | null;
  district_name?: string | null;
  approval_date?: string | null;
  approved_by_name?: string | null;
  remarks?: string | null;
  kpis: SchemeProgressKpis;
  families: AffectedFamilyListItem[];
  allotments: RAndRAllotmentItem[];
  updated_at?: string | null;
}

export interface AcquisitionTraceLinkage {
  project_id: string;
  project_code: string;
  project_title: string;
  parcel_id?: string | null;
  khasra_number?: string | null;
  khata_number?: string | null;
  parcel_area_sqm?: number | string | null;
  parcel_status?: string | null;
  owner_id?: string | null;
  owner_name?: string | null;
  compensation_id?: string | null;
  compensation_reference?: string | null;
  compensation_total_inr?: number | string | null;
  compensation_status?: string | null;
  award_id?: string | null;
  award_number?: string | null;
  award_status?: string | null;
  disbursement_id?: string | null;
  disbursement_reference?: string | null;
  disbursement_status?: string | null;
  disbursement_paid_amount?: number | string | null;
  possession_id?: string | null;
  possession_reference?: string | null;
  possession_status?: string | null;
  possession_date?: string | null;
  scheme_id?: string | null;
  scheme_title?: string | null;
  scheme_reference?: string | null;
}

export interface AffectedFamilyListItem {
  id: string;
  family_reference_id: string;
  head_of_family_name: string;
  family_type: string;
  displacement_category: string;
  social_category: string;
  village_name?: string | null;
  khasra_number?: string | null;
  parcel_id?: string | null;
  project_id?: string | null;
  project_code?: string | null;
  project_title?: string | null;
  scheme_id?: string | null;
  scheme_title?: string | null;
  scheme_reference?: string | null;
  eligibility_status: string;
  eligibility_category?: string | null;
  rehabilitation_status: string;
  allotted_plot_number?: string | null;
  subsistence_grant_inr?: number | string | null;
  is_grant_disbursed: boolean;
  allotments_count: number;
  created_at: string;
}

export interface AffectedFamilyDetail extends AffectedFamilyListItem {
  contact_masked?: string | null;
  family_members_count: number;
  entitled_plot_sqyd?: number | string | null;
  transportation_allowance_inr?: number | string | null;
  one_time_resettlement_allowance_inr?: number | string | null;
  eligibility_assessment_date?: string | null;
  assessing_authority?: string | null;
  eligibility_basis?: string | null;
  eligibility_remarks?: string | null;
  acquisition_trace?: AcquisitionTraceLinkage | null;
  allotments: RAndRAllotmentItem[];
  updated_at?: string | null;
}

export interface RAndRAllotmentItem {
  id: string;
  family_id: string;
  family_reference_id?: string | null;
  head_of_family_name?: string | null;
  scheme_id?: string | null;
  scheme_title?: string | null;
  allotment_reference: string;
  entitlement_category: string;
  allotment_type: string;
  asset_identifier?: string | null;
  allotment_order_no?: string | null;
  allotment_date?: string | null;
  delivery_date?: string | null;
  allocated_value_inr?: number | string | null;
  status: string;
  responsible_authority?: string | null;
  remarks?: string | null;
  created_at: string;
}

export interface RAndRSchemeFilterParams {
  project_id?: string;
  status?: string;
  scheme_type?: string;
  search?: string;
  page?: number;
  page_size?: number;
}

export interface AffectedFamilyFilterParams {
  project_id?: string;
  scheme_id?: string;
  district_id?: string;
  eligibility_status?: string;
  rehabilitation_status?: string;
  search?: string;
  page?: number;
  page_size?: number;
}

export interface EligibilityAssessmentPayload {
  eligibility_status: "PENDING" | "UNDER_REVIEW" | "ELIGIBLE" | "INELIGIBLE" | "DISPUTED" | "APPROVED";
  eligibility_category?: string;
  assessing_authority?: string;
  eligibility_basis?: string;
  eligibility_remarks?: string;
}

export interface RAndRAllotmentPayload {
  family_id: string;
  scheme_id?: string;
  entitlement_category: string;
  allotment_type: "PLOT" | "HOUSING_UNIT" | "CASH_GRANT" | "ANNUITY" | "LIVELIHOOD_ASSET" | "TRAINING_SEAT" | "OTHER";
  asset_identifier?: string;
  allotment_order_no?: string;
  allotment_date?: string;
  delivery_date?: string;
  allocated_value_inr?: number | string;
  status?: string;
  responsible_authority?: string;
  remarks?: string;
}

File: backend/app/schemas/randr.py
import uuid
from datetime import date, datetime
from decimal import Decimal
from typing import List, Optional, Any
from pydantic import BaseModel, ConfigDict, Field


# ---------------------------------------------------------------------------
# Scheme Progress KPIs
# ---------------------------------------------------------------------------
class SchemeProgressKpis(BaseModel):
    total_affected_families: int = 0
    eligible_families: int = 0
    approved_families: int = 0
    allocated_families: int = 0
    completed_families: int = 0
    completion_percent: float = 0.0


# ---------------------------------------------------------------------------
# R&R Scheme Schemas
# ---------------------------------------------------------------------------
class RAndRSchemeBase(BaseModel):
    project_id: uuid.UUID
    scheme_title: str = Field(..., max_length=200)
    scheme_type: str = Field("RESETTLEMENT_COLONY", max_length=50)
    resettlement_site_name: str = Field(..., max_length=200)
    total_plots_planned: int = 0
    sanctioned_budget_cr: Decimal = Decimal("0.0")
    target_completion_date: Optional[date] = None
    remarks: Optional[str] = None


class RAndRSchemeCreate(RAndRSchemeBase):
    scheme_reference: Optional[str] = None


class RAndRSchemeUpdate(BaseModel):
    scheme_title: Optional[str] = None
    scheme_type: Optional[str] = None
    resettlement_site_name: Optional[str] = None
    total_plots_planned: Optional[int] = None
    sanctioned_budget_cr: Optional[Decimal] = None
    spent_budget_cr: Optional[Decimal] = None
    status: Optional[str] = None
    target_completion_date: Optional[date] = None
    remarks: Optional[str] = None


class RAndRSchemeStatusUpdate(BaseModel):
    status: str = Field(..., description="DRAFT, UNDER_REVIEW, APPROVED, ACTIVE, COMPLETED, ON_HOLD, CANCELLED")
    remarks: Optional[str] = None


class RAndRSchemeListItem(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    project_id: uuid.UUID
    project_code: Optional[str] = None
    project_title: Optional[str] = None
    scheme_reference: Optional[str] = None
    scheme_title: str
    scheme_type: str
    resettlement_site_name: str
    state_name: Optional[str] = None
    district_name: Optional[str] = None
    total_plots_planned: int
    total_plots_allotted: int
    sanctioned_budget_cr: Decimal
    spent_budget_cr: Decimal
    status: str
    target_completion_date: Optional[date] = None
    total_families_count: int = 0
    eligible_families_count: int = 0
    assisted_families_count: int = 0
    progress_percent: float = 0.0
    created_at: Optional[datetime] = None


# ---------------------------------------------------------------------------
# Acquisition Linkage Trace
# ---------------------------------------------------------------------------
class AcquisitionTraceLinkage(BaseModel):
    project_id: uuid.UUID
    project_code: str
    project_title: str
    parcel_id: Optional[uuid.UUID] = None
    khasra_number: Optional[str] = None
    khata_number: Optional[str] = None
    parcel_area_sqm: Optional[Decimal] = None
    parcel_status: Optional[str] = None
    owner_id: Optional[uuid.UUID] = None
    owner_name: Optional[str] = None
    compensation_id: Optional[uuid.UUID] = None
    compensation_reference: Optional[str] = None
    compensation_total_inr: Optional[Decimal] = None
    compensation_status: Optional[str] = None
    award_id: Optional[uuid.UUID] = None
    award_number: Optional[str] = None
    award_status: Optional[str] = None
    disbursement_id: Optional[uuid.UUID] = None
    disbursement_reference: Optional[str] = None
    disbursement_status: Optional[str] = None
    disbursement_amount_inr: Optional[Decimal] = None
    possession_id: Optional[uuid.UUID] = None
    possession_reference: Optional[str] = None
    possession_status: Optional[str] = None


# ---------------------------------------------------------------------------
# R&R Allotment Schemas
# ---------------------------------------------------------------------------
class RAndRAllotmentBase(BaseModel):
    family_id: uuid.UUID
    scheme_id: Optional[uuid.UUID] = None
    allotment_type: str = Field("PLOT", max_length=50)
    entitlement_category: str = Field("HOUSING_RESETTLEMENT", max_length=50)
    asset_identifier: str = Field(..., max_length=100)
    allotment_order_no: str = Field(..., max_length=100)
    allotment_date: date
    delivery_date: Optional[date] = None
    allocated_value_inr: Decimal = Decimal("0.0")
    responsible_authority: Optional[str] = None
    status: str = "ALLOTTED"
    remarks: Optional[str] = None


class RAndRAllotmentCreate(RAndRAllotmentBase):
    allotment_reference: Optional[str] = None


class RAndRAllotmentUpdate(BaseModel):
    status: Optional[str] = None
    delivery_date: Optional[date] = None
    remarks: Optional[str] = None


class RAndRAllotmentItem(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    family_id: uuid.UUID
    family_reference_id: Optional[str] = None
    head_of_family_name: Optional[str] = None
    scheme_id: Optional[uuid.UUID] = None
    scheme_title: Optional[str] = None
    allotment_reference: Optional[str] = None
    entitlement_category: str
    allotment_type: str
    asset_identifier: str
    allotment_order_no: str
    allotment_date: date
    delivery_date: Optional[date] = None
    allocated_value_inr: Decimal
    responsible_authority: Optional[str] = None
    status: str
    remarks: Optional[str] = None
    created_at: Optional[datetime] = None


# ---------------------------------------------------------------------------
# Affected Family Schemas
# ---------------------------------------------------------------------------
class AffectedFamilyBase(BaseModel):
    scheme_id: uuid.UUID
    parcel_id: Optional[uuid.UUID] = None
    land_owner_id: Optional[uuid.UUID] = None
    head_of_family_name: str = Field(..., max_length=100)
    village_name: Optional[str] = None
    family_type: str = Field("PAF_AFFECTED_ONLY", max_length=30)
    displacement_category: str = Field("TITLEHOLDER_DISPLACED", max_length=50)
    social_category: str = Field("GEN", max_length=20)
    family_members_count: int = 4
    contact_masked: Optional[str] = None
    entitled_plot_sqyd: Decimal = Decimal("0.0")
    allotted_plot_number: Optional[str] = None
    subsistence_grant_inr: Decimal = Decimal("0.0")
    transportation_allowance_inr: Decimal = Decimal("0.0")
    one_time_resettlement_allowance_inr: Decimal = Decimal("0.0")
    is_grant_disbursed: bool = False
    rehabilitation_status: str = "SURVEYED"


class AffectedFamilyCreate(AffectedFamilyBase):
    family_reference_id: Optional[str] = None
    eligibility_status: str = "PENDING"
    eligibility_category: Optional[str] = None
    eligibility_assessment_date: Optional[date] = None
    assessing_authority: Optional[str] = None
    eligibility_basis: Optional[str] = None
    eligibility_remarks: Optional[str] = None


class AffectedFamilyUpdate(BaseModel):
    head_of_family_name: Optional[str] = None
    village_name: Optional[str] = None
    family_type: Optional[str] = None
    displacement_category: Optional[str] = None
    social_category: Optional[str] = None
    family_members_count: Optional[int] = None
    contact_masked: Optional[str] = None
    entitled_plot_sqyd: Optional[Decimal] = None
    allotted_plot_number: Optional[str] = None
    subsistence_grant_inr: Optional[Decimal] = None
    transportation_allowance_inr: Optional[Decimal] = None
    one_time_resettlement_allowance_inr: Optional[Decimal] = None
    is_grant_disbursed: Optional[bool] = None
    rehabilitation_status: Optional[str] = None


class EligibilityAssessmentUpdate(BaseModel):
    eligibility_status: str = Field(..., description="PENDING, UNDER_REVIEW, ELIGIBLE, INELIGIBLE, DISPUTED, APPROVED")
    eligibility_category: Optional[str] = None
    eligibility_assessment_date: Optional[date] = None
    assessing_authority: Optional[str] = None
    eligibility_basis: Optional[str] = None
    eligibility_remarks: Optional[str] = None


class RehabilitationStatusUpdate(BaseModel):
    rehabilitation_status: str = Field(..., description="SURVEYED, SCHEME_APPROVED, PLOT_ALLOTTED, SETTLED")
    remarks: Optional[str] = None


class AffectedFamilyListItem(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    family_reference_id: Optional[str] = None
    head_of_family_name: str
    family_type: str
    displacement_category: str
    social_category: str
    village_name: Optional[str] = None
    khasra_number: Optional[str] = None
    parcel_id: Optional[uuid.UUID] = None
    project_id: uuid.UUID
    project_code: Optional[str] = None
    project_title: Optional[str] = None
    scheme_id: uuid.UUID
    scheme_title: Optional[str] = None
    scheme_reference: Optional[str] = None
    eligibility_status: str
    eligibility_category: Optional[str] = None
    rehabilitation_status: str
    allotted_plot_number: Optional[str] = None
    subsistence_grant_inr: Decimal
    is_grant_disbursed: bool
    allotments_count: int = 0
    created_at: Optional[datetime] = None


class AffectedFamilyDetail(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    family_reference_id: Optional[str] = None
    scheme_id: uuid.UUID
    scheme_title: Optional[str] = None
    scheme_reference: Optional[str] = None
    resettlement_site_name: Optional[str] = None
    parcel_id: Optional[uuid.UUID] = None
    land_owner_id: Optional[uuid.UUID] = None
    head_of_family_name: str
    village_name: Optional[str] = None
    family_type: str
    displacement_category: str
    social_category: str
    family_members_count: int
    contact_masked: Optional[str] = None

    # Entitlement details
    entitled_plot_sqyd: Decimal
    allotted_plot_number: Optional[str] = None
    subsistence_grant_inr: Decimal
    transportation_allowance_inr: Decimal
    one_time_resettlement_allowance_inr: Decimal
    is_grant_disbursed: bool

    # Configurable Eligibility Details
    eligibility_status: str
    eligibility_category: Optional[str] = None
    eligibility_assessment_date: Optional[date] = None
    assessing_authority: Optional[str] = None
    eligibility_basis: Optional[str] = None
    eligibility_remarks: Optional[str] = None

    # Lifecycle Status
    rehabilitation_status: str

    # 360 Acquisition Traceability Linkage
    acquisition_trace: Optional[AcquisitionTraceLinkage] = None

    # Allotments received
    allotments: List[RAndRAllotmentItem] = []
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None


# ---------------------------------------------------------------------------
# R&R Scheme 360 Detail
# ---------------------------------------------------------------------------
class RAndRSchemeDetail(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    project_id: uuid.UUID
    project_code: str
    project_title: str
    scheme_reference: Optional[str] = None
    scheme_title: str
    scheme_type: str
    resettlement_site_name: str
    state_name: Optional[str] = None
    district_name: Optional[str] = None
    total_plots_planned: int
    total_plots_allotted: int
    sanctioned_budget_cr: Decimal
    spent_budget_cr: Decimal
    status: str
    target_completion_date: Optional[date] = None
    approval_date: Optional[date] = None
    approved_by_user: Optional[str] = None
    remarks: Optional[str] = None

    # Progress KPIs
    kpis: SchemeProgressKpis

    # Covered families
    families: List[AffectedFamilyListItem] = []
    allotments: List[RAndRAllotmentItem] = []
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None
